google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -2006,7 +2034,19 @@ PooledSessionFuture getSession() throws SpannerException {
                 resourceNotFoundException.getMessage()),
             resourceNotFoundException);
       }
-      sess = sessions.poll();
+      moveExpiredWriteSessionsToSessions();
+      if (wantRwTxnId) {
+        sess = writeSessions.poll();
```

Comment:
I think that we would need to add some more logic here to prevent possible starvation in the following scenario (and similar scenarios):
1. Assume that `maxSessions=10`
2. The client executes 10 read/write transactions in parallel (or enough in parallel to use all 10 different sessions for a r/w transaction). All these sessions have now received a transaction id from the backend, and will be added to the queue of writeSessions and not to the list of sessions.
3. The client wants to execute a read-only transaction, but there are no sessions in the normal session list, and `maxSessions` has been reached. It will therefore wait indefinitely.

Also, this list of writeSessions does not seem to be considered when calculating the number of sessions currently in the pool, is not considered for sessions that should be removed from the pool because they have expired, and is not included in the logic for closing the session pool and reporting any leaked sessions. (I know it's prototype so it's logical that not all of this is implemented, but my point is that more complexity than this will still be needed to be added to the session pool.)